examples/05_tool_inference_lock/ex_tool_inference_lock/inference_lock_helper.rb

```diff
@@ -0,0 +1,122 @@
+module Examples
+  module ToolInferenceLock
+    # Helper object to lock tool inference using Shift or arrow keys.
+    class InferenceLockHelper
+      # Create InferenceLockHelper object.
+      #
+      # @para tool [Object] The SketchUp Ruby tool using this helper object.
+      # @param active_ip [Sketchup::InputPoint]
+      # @param reference_ip [Sketchup::InputPoint]
+      def initialize(tool, active_ip, reference_ip = active_ip)
+        @tool = tool
+        @active_ip = active_ip
+        @reference_ip = reference_ip
+
+        @axis_lock = nil
+        @mouse_x = nil
+        @mouse_y = nil
+      end
+
+      # Call this method from +onKeyDown+.
+      def handle_keydown(key, view)
+        if key == CONSTRAIN_MODIFIER_KEY
+          try_lock_constraint(view)
+        else
+          try_lock_axis(key, view)
+        end
+
+        # Emulate mouse move to update InputPoint picking.
+        @tool.onMouseMove(0, @mouse_x, @mouse_y, view) if @mouse_x && @mouse_y
```

Comment:
This is a strange coupling. I need to read the rest of the code before I can make a suggestion here.

Reply:
As per comment further up, I think the InputPoint should be owned by this class. And the onMouseMove logic in the tool (apart from invalidating the view and updating the VCB) should be done in this class.

If we do that, then we can from `handle_keydown` call this instance's `handle_mousemove` and we eliminate the need for a backreference to the tool - further decoupling logic. Child object knowing and delegating back to parent (owning) object directly is a smell.

Reply:
To provide some more example of separation of responsibility; the TrueBend tool had a lot of logic and state, screen input widgets, input text, the bending logic, the drawing logic. But they are split out into separate classes and the tool itself is as slim as it can be - acting only as the glue between everything (a Controller).

https://github.com/thomthom/true-bend/blob/master/src/tt_truebend/tools/truebend.rb

I even did a class for parsing the VCB input (it can accept a number of inputs determined by tool state). And that parser class then emits custom errors which I'm catching that completely isolate it from generic `ArgumentError` etc.

I'm not saying we should do all of that in this example. But just wanted to provide an example of what I mean by separation of responsibility. Using events is a key technique in this - for instance the manipulator widget uses events that the owning tool can use - without it ever needing to know about the tool itself; 
https://github.com/thomthom/true-bend/blob/65c1c9a8a81a81abfdf185915e178d2ee8bb5c9a/src/tt_truebend/tools/truebend.rb#L41-L47

Reply:
I've wanted to give a lot of the inputpoint control to the individual tool. Some tools only have a single picking inputpoint (and possibly one as reference), some tools have a whole series of input points and changes which one is currently picking depending on tool state. What style you use seems to both depend on the tool and be a personal preference. I've never been entirely happy about emulating the mouse move, but it lets the tool handle how it uses inputpoints.

What would you think about letting the tool own the inputpoints completely and just pass them along as arguments when the InferenceLockHelper needs them?

Reply:
How does it look to you know? I made the tool own the InputPoints as they are quite closely integrated to the tool's logic. The InferenceLockHelper class is really a wrapper to functionality of the View object, not the InputPoints.

Reply:
Yea, I'm also ok with the tool owning them and passing them along. Key thing is that shared ownership/knowledge should be avoided.

Reply:
But I'm not that keen on this direct call back to the tool. This should ideally be decoupled with an event or notification.